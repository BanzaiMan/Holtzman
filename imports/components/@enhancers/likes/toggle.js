
// @flow

// $FlowMeteor
import { Meteor } from "meteor/meteor";
import React, { Component } from "react";
import { graphql } from "react-apollo";
import gql from "graphql-tag";
import { connect } from "react-redux";

import {
  nav as navActions,
  liked as likedActions,
  modal
} from "../../../data/store";

import OnBoard from "../../people/accounts";

const TOGGLE_LIKE_MUTATION = gql`
  mutation ToggleLike($nodeId: String!) {
    toggleLike(nodeId: $nodeId) {
      like {
        id
      }
    }
  }
`;

const withToggleLike = graphql(TOGGLE_LIKE_MUTATION, {});

export const classWrapper = (propsReducer: Function) => (WrappedComponent: any) => {
  type ILikesWrapper = {
    dispatch: Function,
    mutate: Function,
  };

  class LikesWrapper extends Component {
    props: ILikesWrapper;

    componentWillMount() {
      this.props.dispatch(navActions.setLevel("CONTENT"));
      this.props.dispatch(navActions.setAction("CONTENT", {
        id: 2,
        action: this.toggleLike,
      }));
    }

    getNodeId = () => {
      if (propsReducer && typeof propsReducer === "function") {
        return propsReducer(this.props);
      }
      console.warn("propsReducer was either not passed, or is not a function");
      return null;
    };

    toggleLike = () => {
      const { dispatch, mutate } = this.props;
      if (!Meteor.userId()) { // if not logged in, show login modal
        dispatch(modal.render(OnBoard, {
          coverHeader: true, modalBackground: "light",
        }));
      } else { // if logged in, toggle like state in redux, remote with gql query
        const nodeId = this.getNodeId();
        if (nodeId) {
          dispatch(likedActions.toggle({ entryId: nodeId }));
          mutate({ variables: { nodeId } });
        }
      }

      return { type: "FALSY", payload: {} };
    }

    render = () => <WrappedComponent {...this.props} onLike={this.toggleLike} />;
  }
  return connect()(withToggleLike(LikesWrapper));
};

export default classWrapper;
