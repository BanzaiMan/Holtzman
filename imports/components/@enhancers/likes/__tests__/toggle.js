
import { classWrapper } from "../toggle";

describe("Likes Wrapper", () => {

  beforeEach(() => {

  });

  it("", () => {

  });

});
