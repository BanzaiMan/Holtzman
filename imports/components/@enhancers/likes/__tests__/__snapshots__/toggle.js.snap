exports[`Likes Wrapper should contain mutation 1`] = `
"mutation ToggleLike($nodeId: String!) {
  toggleLike(nodeId: $nodeId) {
    like {
      id
    }
  }
}
"
`;

exports[`Likes Wrapper should render the child component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }>
  <ApolloProvider
    client={
      Object {
        "initStore": [Function],
        "mutate": [Function],
      }
    }
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "subscribe": [Function],
      }
    }>
    <Connect(Apollo(LikesWrapper))>
      <Apollo(LikesWrapper)
        dispatch={[Function]}>
        <LikesWrapper
          dispatch={[Function]}
          mutate={[Function]}>
          <TestComponent
            dispatch={[Function]}
            mutate={[Function]}>
            <div
              id="tester">
              Test
            </div>
          </TestComponent>
        </LikesWrapper>
      </Apollo(LikesWrapper)>
    </Connect(Apollo(LikesWrapper))>
  </ApolloProvider>
</Provider>
`;
